keps/sig-storage/3751-volume-attributes-class/README.md

```diff
@@ -435,6 +438,50 @@ Operation metrics from [csiOperationsLatencyMetric](https://github.com/kubernete
 
 ## Design Details
 
+#### Binding of PV and PVC
+
+Creating the bidirectional binding between a PV and PVC is delicate, because
+there is no transaction support in the kubernetes API to do the whole thing
+atomically. Binding with volume attributes adds to this process but does not
+fundamentally change the 4 steps performed. To support the VolumeAttributesClass
+the binding process will become the following.
+
+##### Find a PV matching the PVC
+
+If the PVC is pre-bound to a PV, the process is unchanged: check if node
```

Comment:
There's actually two kinds of prebinding:
1. PV.spec.claimRef. Here we assume the admin knows what they're doing and we don't check storageclass matches: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go#L247C1-L248C1
2.  PVC.spec.volume. This is user requested and we do check that storageclass matches: https://github.com/kubernetes/kubernetes/blob/c49b140c45e0e9b7e47da4bbe81d32829008c06d/pkg/controller/volume/persistentvolume/pv_controller.go#L421

I think in 1), we need to consider the ramifications of if the admin binds something and the VAC doesn't match. The binding will be successful but we will start to try to modify the volume since the VACs don't match.

Reply:
In neither case do we propose to verify matching VAC. If they don't match, then after binding it will look to the resizer like a VAC change on the PVC and it will try to reconcile the PV to match. If for some reason the VAC on the PVC is invalid, there will be error messages, but the volume will still be usable under the PV's VAC. 

This seems the right semantics to me. The alternative would be to block binding until the PVC is manually updated which I think would lead to a very fragile user experience.

Reply:
I think there's a potential quota implication though. Say the user only has quota for slow-vac, but they request a PV with fast-vac (and lets also say it's impossible to change from fast-vac to slow-vac). Quota only looks at PVC.spec.vac during admission so will allow it. Now they've gotten fast-vac and because conversion to slow-vac is not supported, they will continue to use fast-vac.

Reply:
Discussing offline, there is also a race if someone creates a PVC with VAC and then changes the VAC quickly before we bound the PV. To handle that race, I think we need to enforce that the VAC cannot be modified until the PVC is bound.

cc @carlory @jsafrane

Reply:
+1 @msau42 's convinced me that quota-hacking means we have to enforce matching VAC, and since there's no was to undo the binding process once started, it means we have to prevent VAC from changing during this process.

Reply:
During the creation process, if a pvc references a unavaliable vac which contains unsupported attributes by csi driver, the pv won't be created. If we don't allow users to change the vac, users or external operators have to delete and re-create it with new vac. Is it desired behavior?

Reply:
After some discussion, @msau42 and I came to the conclusion that was ok. Forcing delete & recreation saves us from many race conditions. It doesn't seem to affect the UX much as create-update vs create-delete-update before the PVC has been used seems about the same.

I think forcing delete & recreate is ultimately easier to reason about for a user anyway -- the race conditions we have to worry about, they would have to worry about eventually too.

Reply:
Okay, I agree with your opinion. In other words, is it forbidden to update the desired vac of a PVC from nil/empty value or other value unless the PVC is in the `Bound` state?

If so, we must change the existing validation logic when updating a PVC. I can include the change in https://github.com/kubernetes/kubernetes/pull/121902.

Reply:
Yes, I think that's true.

I just pushed an update with the proposed change. PTAL.

Do you know where the source is for png diagrams? I think we should update them to make the new mutation validation clear.

Note that the KEP already said that "modify volume is only allowed on bound PVCs". Does this mean that we already prevented volume capacity from being changed before the volume was bound? or we just ignored it until it was bound?

Reply:
At least looking at the [api validation logic](https://github.com/kubernetes/kubernetes/blob/6c556cb9c22d3f4686cbea41e6d301ba889fd6a6/pkg/apis/core/validation/validation.go#L2389-L2403), we don't enforce it, so there's a bug there. Need to check resizer code.

Reply:
@sunnylovestiramisu didn't upload the source file. You may have to redraw the diagram.

Reply:
Oh, so we could validate VAC mutable-only-when-bound in api validation rather than a webhook? That's a good point.

Reply:
I think the code does prevent changes to capacity when unless the PVC is bound: if it's bound, the capacity value is copied over from old to new, so that the DeepEqual will be true; if it's not bound it's not copied, so if it's changed the DeepEqual will fail. An overly-clever way to do it IMHO.

(those comments "let's make sure * is the same" are also very confusing to me. Those assignments actually allow those fields to **not** be the same)...

Reply:
> @sunnylovestiramisu didn't upload the source file. You may have to redraw the diagram.

:-P I'll razz her about that when she's back from leave ;-)